packages/core/crates/core-packet/src/interaction.rs

```diff
@@ -1569,34 +1574,45 @@ mod tests {
         };
 
         // Check that we received all acknowledgements and packets
-        let finish = async move {
-            let (mut acks, mut pkts) = (0, 0);
-            for _ in 0..2 * PENDING_PACKETS {
-                match select(ack_rx.next(), pkt_rx.next()).await {
-                    Either::Left((_, _)) => {
-                        acks += 1;
-                    }
-                    Either::Right((pkt, _)) => {
-                        let msg = pkt.unwrap();
-                        assert_eq!(test_msg, msg, "received packet payload must match");
-                        pkts += 1;
-                    }
-                }
-            }
-            (acks, pkts)
-        };
-
+        let finish = join3(
+            ack_rx.collect::<Vec<_>>(),
+            ack_tkt_rx.collect::<Vec<_>>(),
+            pkt_rx.collect::<Vec<_>>(),
+        );
         let timeout = async_std::task::sleep(Duration::from_secs(TIMEOUT_SECONDS));
         pin_mut!(finish, timeout);
 
-        let succeeded = match select(finish, timeout).await {
-            Either::Left(((acks, pkts), _)) => {
-                assert_eq!(acks, PENDING_PACKETS, "did not receive all acknowledgements");
-                assert_eq!(pkts, PENDING_PACKETS, "did not receive all packets");
-                succeeded && true
-            }
-            Either::Right(_) => false,
-        };
+        let succeeded = succeeded
+            && match select(finish, timeout).await {
+                Either::Left(((acks, ack_tkts, pkts), _)) => {
+                    debug!(
+                        "acks: {}, ack_tkts: {}, pkts: {}",
+                        acks.len(),
+                        ack_tkts.len(),
+                        pkts.len()
+                    );
+                    assert_eq!(acks.len(), PENDING_PACKETS, "did not receive all acknowledgements");
```

Comment:
It'd be more descriptive to have the debug output string inside the assert_eq! macro so that it is output on fail, but not on success.